### DISCUSSION TITLE
Running Hermes for VisionOS

### DISCUSSION THREAD

**Discussion by okwasniewski** [facebook/hermes#1164] [General]
Hey! 

I'm working on bringing React Native to VisionOS ([here](https://github.com/callstack/react-native-visionos)). I've been trying to compile the app with Hermes enabled but unfortunately without success. 

After bringing this topic on RN Contributors Discord folks from Microsoft working on RN MacOS pointed out that building Hermes from source didn't work for them either and they suggested trying out the prebuilt version. 

However, the prebuilt version doesn't support VisionOS (as it supports MacOS and iOS). I don't want to follow the path of React Native TvOS which has its own fork of Hermes. Can we add support for VisionOS in Hermes podspec or is there any different way to make it work for Out Of Tree Platforms? 

Coming back to the error I was getting when building from source: `Undefined symbols: hermes::vm::NopCrashManager::~NopCrashManager()`

<details>
  <summary>Expand the full error log</summary>

```
ld: Undefined symbols:
  hermes::vm::NopCrashManager::~NopCrashManager(), referenced from:
      std::__1::default_delete<hermes::vm::NopCrashManager>::operator()[abi:v160006](hermes::vm::NopCrashManager*) const in libReact-hermes.a[5](HermesExecutorFactory.o)
  facebook::hermes::inspector_modern::chrome::CDPHandler::registerCallbacks(std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)>, std::__1::function<void ()>), referenced from:
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::addPage(std::__1::shared_ptr<facebook::hermes::inspector_modern::chrome::CDPHandler>)::$_0::operator()(std::__1::unique_ptr<facebook::react::jsinspector_modern::IRemoteConnection, std::__1::default_delete<facebook::react::jsinspector_modern::IRemoteConnection>>) const in libReact-hermes.a[4](ConnectionDemux.o)
  facebook::hermes::inspector_modern::chrome::CDPHandler::unregisterCallbacks(), referenced from:
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::removePage(int) in libReact-hermes.a[4](ConnectionDemux.o)
      facebook::hermes::inspector_modern::chrome::(anonymous namespace)::LocalConnection::disconnect() in libReact-hermes.a[4](ConnectionDemux.o)
  facebook::hermes::inspector_modern::chrome::CDPHandler::create(std::__1::unique_ptr<facebook::hermes::inspector_modern::RuntimeAdapter, std::__1::default_delete<facebook::hermes::inspector_modern::RuntimeAdapter>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool), referenced from:
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::enableDebugging(std::__1::unique_ptr<facebook::hermes::inspector_modern::RuntimeAdapter, std::__1::default_delete<facebook::hermes::inspector_modern::RuntimeAdapter>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libReact-hermes.a[4](ConnectionDemux.o)
  facebook::hermes::inspector_modern::chrome::CDPHandler::handle(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>), referenced from:
      facebook::hermes::inspector_modern::chrome::(anonymous namespace)::LocalConnection::sendMessage(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) in libReact-hermes.a[4](ConnectionDemux.o)
  facebook::hermes::makeHermesRuntime(hermes::vm::RuntimeConfig const&), referenced from:
      facebook::react::HermesExecutorFactory::createJSExecutor(std::__1::shared_ptr<facebook::react::ExecutorDelegate>, std::__1::shared_ptr<facebook::react::MessageQueueThread>) in libReact-hermes.a[5](HermesExecutorFactory.o)
  facebook::hermes::inspector_modern::chrome::CDPHandler::getTitle() const, referenced from:
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::enableDebugging(std::__1::unique_ptr<facebook::hermes::inspector_modern::RuntimeAdapter, std::__1::default_delete<facebook::hermes::inspector_modern::RuntimeAdapter>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) in libReact-hermes.a[4](ConnectionDemux.o)
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::removePage(int) in libReact-hermes.a[4](ConnectionDemux.o)
      facebook::hermes::inspector_modern::chrome::ConnectionDemux::addPage(std::__1::shared_ptr<facebook::hermes::inspector_modern::chrome::CDPHandler>) in libReact-hermes.a[4](ConnectionDemux.o)
      facebook::hermes::inspector_modern::chrome::(anonymous namespace)::LocalConnection::LocalConnection(std::__1::shared_ptr<facebook::hermes::inspector_modern::chrome::CDPHandler>, std::__1::shared_ptr<std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>>) in libReact-hermes.a[4](ConnectionDemux.o)
      facebook::hermes::inspector_modern::chrome::(anonymous namespace)::LocalConnection::disconnect() in libReact-hermes.a[4](ConnectionDemux.o)
  vtable for hermes::vm::CrashManager, referenced from:
      hermes::vm::CrashManager::CrashManager() in libReact-hermes.a[5](HermesExecutorFactory.o)
  vtable for hermes::vm::NopCrashManager, referenced from:
      hermes::vm::NopCrashManager::NopCrashManager() in libReact-hermes.a[5](HermesExecutorFactory.o)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
 
</details>

Feel free to let me know if you need more details!

**Comment by Saadnajmi** [comment]
Just to chime in, we saw this error building Hermes from source for RN macOS too. Looks like something broke recently on main.

**Comment by tmikov** [comment]
We will look into this. FWIW, I am building Hermes on MacOS daily and it certainly builds without error. However I am not using CocoaPods. Unfortunately no one on our team really understands or uses CocoaPods. Last time I tried even installing it, it was a challenge (see https://x.com/tmikov/status/1705991622100676812?s=20). 

FWIW, I don't understand why TvOS needs a fork of Hermes. IIRC, there was a PR which needed some minor improvements, but they didn't seem interested in working on it.

**Reply by Saadnajmi** [reply]
@shwanton might have more context on the macOS + cocoapods hit

**Reply by okwasniewski** [reply]
Okay I will try to open a PR to support VisionOS based on the PR created for tvOS: https://github.com/facebook/hermes/pull/1000

**Comment by neildhar** [comment]
Hey @okwasniewski, could you share repro steps for observing this linker error on macOS? These symbols are exported by libhermes, so they should be resolved when you link against libhermes.

**Reply by okwasniewski** [reply]
Hey! 

So as @Saadnajmi pointed out in the commend above the issue occurs when you try to build RN MacOS repo together with building Hermes from source (same with building VisionOS). 

In order to reproduce this you would need to clone `react-native-macos` repo: https://github.com/microsoft/react-native-macos
and then change `hermes_enabled = true` [here](https://github.com/microsoft/react-native-macos/blob/main/packages/rn-tester/Podfile) and build the app by following the README steps.

**Reply by tmikov** [reply]
Hi @okwasniewski, I suspect that we can't really provide a lot of meaningful help for Microsoft's RN fork. Not because we don't want to, we are just not familiar with it. Are you trying to use that fork in combination with mainline Hermes? Is that something that is supported by Microsoft, or do they require using their own fork of ~~Windows~~Hermes?

**Reply by Saadnajmi** [reply]
I should note that the Hermes break we saw was reported by someone on Meta's messenger desktop team. React Native macOS doesn't do anything custom to Hermes, we have zero diffs and enable it with the same `USE_HERMES pod install` as upstream React Native. The difference is:
1) We can't use the prebuilt binaries that React Native uses, as our version numbers don't match
2) We don't use Hermes for macOS internally at the moment, so our CI doesn't check if Hermes + React Native macOS breaks. 

The Hermes build issue should not at all be related to the fact we have a macOS fork for React Native.

**Reply by neildhar** [reply]
@Saadnajmi thank you for sharing that context.

I think we'll need help from someone familiar with the RN build for macOS in investigating this, as our expertise here is very limited. The attached linker error is not occurring when building Hermes itself from source, but rather when linking the RN components that depend on Hermes. Is it possible that these targets are not being linked against the Hermes framework for some reason?

**Reply by Saadnajmi** [reply]
Curious, how can an RN component depend on Hermes? I thought in general, RN components would be JS engine agnostic. For someone familiar, @shwanton should be able to help

**Reply by neildhar** [reply]
The failure here is in HermesExecutorFactory and in the parts of the Hermes inspector that live on the RN side. These take direct dependencies on functions exported by the Hermes framework.

**Reply by Saadnajmi** [reply]
Ah, so _native_ RN components. I thought you meant like, a JS react component. Makes sense!

**Reply by dannysu** [reply]
Does the same issue occur when not building for Debug?
In the upstream RN repo, there are various places that check for `HERMES_ENABLE_DEBUGGER`, which should be true only for Debug.

**Reply by neildhar** [reply]
@okwasniewski Given what @Saadnajmi said above, a good starting point for investigating this may be to bisect the problem in both Hermes/RN and identify when/what change caused this build to stop working.

If you could share the results of your bisect, we can help figure out what to do next.

**Reply by okwasniewski** [reply]
Thanks for the insights! I will check it out and provide details of my investigation here.

**Comment by okwasniewski** [comment]
I've spent a significant amount of time debugging what's actually happening here. And it turns out for MacOS Hermes wasn't properly linked for local Xcode builds. I've adjusted the build script here to fix the issue for MacOS: https://github.com/facebook/react-native/pull/41387

Regarding visionOS first issue was the same - Hermes wasn't linked properly. But it also wasn't built targeting visionOS (it was built for MacOS as it's the default for CMake). 

The issue was laying inside of CMake and flag that it passes (`-mtargetos=xros`) it was passing it for both `XROS` and `XRSimulator` builds. And it wasn't properly building the app. 

Only after temporarily passing correct target flag (and compiling CMake from source removing previous flags), it started to work: 

```
if(HERMES_APPLE_TARGET_PLATFORM MATCHES "xrsimulator")
  set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} "-target arm64-apple-xros1.0-simulator")
endif()
if(HERMES_APPLE_TARGET_PLATFORM MATCHES "xros")
  set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} "-target arm64-apple-xros1.0")
endif()
```

Here is the corresponding issue in CMake repo: https://gitlab.kitware.com/cmake/cmake/-/issues/25188

I'll wait with adding visionOS support to Hermes until that issue is resolved and a stable version of CMake 3.28 will be released. 

Screenshot of visionOS running Hermes: 

![image](https://github.com/facebook/hermes/assets/52801365/cdb6e714-8dfb-49a4-8988-0b4ad13c7cf1)
